api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,629 @@
+		}
+
+		if len(deployIDCopies) == 0 {
+			continue
+		}
+
+		// iterate through deploys
+		for deployID := range deployIDCopies {
+			c, err := e.getContainerResourceFromDeploy(deployID)
+			if err != nil {
+				return nil, err
+			}
+
+			resourceConsumers = append(resourceConsumers, c...)
+		}
+	}
+
+	// GET PENDING TASK RESOURCE CONSUMERS IN ECS
+	taskSummaries, err := e.TaskProvider.List()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, summary := range taskSummaries {
+		if summary.EnvironmentID == clusterName {
+
+			task, err := e.TaskProvider.Read(summary.TaskID)
+			if err != nil {
+				return nil, err
+			}
+
+			taskResourceConsumers, err := e.getContainerResourceFromDeploy(task.DeployID)
+			if err != nil {
+				return nil, err
+			}
+
+			resourceConsumers = append(resourceConsumers, taskResourceConsumers...)
+		}
+	}
+
+	// GET PENDING TASK RESOURCE CONSUMERS IN JOBS
+	jobs, err := e.JobStore.SelectAll()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, job := range jobs {
+		if job.Type == models.CreateDeployJob {
+			if job.Status == models.PendingJobStatus || job.Status == models.InProgressJobStatus {
+				var req models.CreateTaskRequest
+				if err := json.Unmarshal([]byte(job.Request), &req); err != nil {
+					return nil, err
+				}
+
+				if req.EnvironmentID == clusterName {
+					// note that this isn't exact if the job has started some, but not all of the tasks
+					taskResourceConsumers, err := e.getContainerResourceFromDeploy(req.DeployID)
+					if err != nil {
+						return nil, err
+					}
+
+					resourceConsumers = append(resourceConsumers, taskResourceConsumers...)
+				}
+			}
+		}
+	}
+
+	return resourceConsumers, nil
+}
+
+func (e *EnvironmentScaler) calculateNewProvider(clusterName string) (*models.ResourceProvider, error) {
+	input := &autoscaling.CreateAutoScalingGroupInput{}
+	if err := input.Validate(); err != nil {
+		return nil, err
+	}
+
+	group, err := e.Client.AutoScaling.CreateAutoScalingGroup(input)
+	if err != nil {
+		return nil, err
+	}
+
+	inputLaunch := &autoscaling.DescribeLaunchConfigurationsInput{}
+	inputLaunch.SetLaunchConfigurationNames([]*string{aws.String(group.String())})
+
+	config, err := e.Client.AutoScaling.DescribeLaunchConfigurations(inputLaunch)
+	if err != nil {
+		return nil, err
+	}
+
+	var instanceType string
+	if len(config.LaunchConfigurations) > 0 {
+		instanceType = aws.StringValue(config.LaunchConfigurations[0].InstanceType)
+	}
+
+	// these ports are automatically used by the ecs agent
+	defaultPorts := []int{
+		22,
+		2376,
+		2375,
+		51678,
+		51679,
+	}
+
+	resource := &models.ResourceProvider{}
+	resource.ID = "<new instance>"
+	resource.InUse = false
+	resource.UsedPorts = defaultPorts
+	resource.AvailableMemory = instanceType
+	return resource, nil
+}
+
+func (e *EnvironmentScaler) getContainerResourceFromDeploy(deployID string) ([]models.ResourceConsumer, error) {
+	// use some kind of deploy cache
+	if consumers, ok := e.deployCache[deployID]; ok {
```

Comment:
See previous comment about thread-safety. The easiest approach I suppose could be to have two instances of scaler (one for scaling and one for the job runner).